### PR TITLE
Don't allow any rate drift when rate is snapped

### PR DESCRIPTION
**Pull request by alex-cardstack**
CS-2883

### CODE REVIEW

File: test/Exchange-test.js [hunk 1]
```diff
@@ -100,5 +104,55 @@ contract("Exchange", (accounts) => {
...
+
+      expect(await daiOracle.isSnappedToUSD()).to.be.ok;
+
+      expect(
+        await exchange.isAllowableRate(
+          daicpxdToken.address,
+          BigNumber.from("100000000")
+        )
+      ).to.be.ok;
+      expect(
```
Comment (tintinthong) [inline test/Exchange-test.js:144]: 👍

File: test/Exchange-test.js [hunk 2]
```diff
@@ -100,5 +104,55 @@ contract("Exchange", (accounts) => {
...
+      expect(await exchange.isAllowableRate(cardcpxdToken.address, "980000")).to
+        .not.be.ok;
+      expect(await exchange.isAllowableRate(cardcpxdToken.address, "1010000"))
+        .to.be.ok;
+      expect(await exchange.isAllowableRate(cardcpxdToken.address, "1020000"))
+        .to.not.be.ok;
+
+      expect(await exchange.isAllowableRate(cardcpxdToken.address, "1005000"))
+        .to.be.ok;
+    });
```
Comment (tintinthong) [inline test/Exchange-test.js:129]:
```
 Actual rate 1000000                                                                                                                                                                                                                                                          
 Requested rate 1005000                                                                                                                                                                                                                                                       
 ObserverDriftPercent 500000                                                                                                                                                                                                                                                  
 RateDriftPercent 1000000  
```

File: test/Exchange-test.js [hunk 3]
```diff
@@ -100,5 +104,55 @@ contract("Exchange", (accounts) => {
...
+
+      expect(await exchange.isAllowableRate(cardcpxdToken.address, "999999")).to
+        .be.ok;
+      expect(await exchange.isAllowableRate(cardcpxdToken.address, "990000")).to
+        .be.ok;
+      expect(await exchange.isAllowableRate(cardcpxdToken.address, "980000")).to
+        .not.be.ok;
+      expect(await exchange.isAllowableRate(cardcpxdToken.address, "1010000"))
+        .to.be.ok;
+      expect(await exchange.isAllowableRate(cardcpxdToken.address, "1020000"))
```
Comment (tintinthong) [inline test/Exchange-test.js:124]:
```
 Actual rate 1000000                                                                                                                                                                                                                                                          
 Requested rate 1020000                                                                                                                                                                                                                                                       
 ObserverDriftPercent 2000000                                                                                                                                                                                                                                                 
 RateDriftPercent 1000000 
```

Comment (tintinthong) [reply test/Exchange-test.js:124]: `1000000 /10^8=0.01 = 1%`?

**Review by tintinthong — APPROVED** [review]
👍

File: test/Exchange-test.js [hunk 4]
```diff
@@ -100,5 +104,55 @@ contract("Exchange", (accounts) => {
         .convertFromCARD(daicpxdToken.address, toTokenUnit(100))
         .should.be.rejectedWith(Error, "no exchange exists for token");
     });
+
+    it("allows rate drift less than rate drift percentage", async () => {
+      let rate = BigNumber.from("1000000");
+      let driftPercentage = await exchange.rateDriftPercentage();
+      // assumptions in this test only valid for 10% rate drift
```
[the remark this answers is not in the document]
Comment (alex-cardstack) [reply test/Exchange-test.js]: Yes that's correct, updated

**Pull request merged**